### DISCUSSION TITLE
What is the most efficient way to move files between different datasets in the same zpool?

### DISCUSSION THREAD

**Discussion by shivabohemian** [openzfs/zfs#15447] [Q&A] [answered]
Hello, 
When I tried to move large files or folders between different datasets in the same zpool, I found it slow to use the `mv` command. Is there a more efficient way?

**Comment by IvanVolosyuk** [comment, marked as answer]
The 'mv' command between different datasets behaves like 'cp'. That's the reason why it is slow. On the other hand, in newly released ZFS 2.2 there is reflink support, so technically ```cp --reflink=auto src dst``` should be instant and you can delete the source file after that.

**Reply by shivabohemian** [reply]
Thank you for your reply, I'll try it.

**Reply by shivabohemian** [reply]
Hello, I've tried `cp --reflink=auto src dst` in released ZFS 2.2. It works in the same dataset but reports an error `cp: failed to clone xxx from yyy: Invalid cross-device link` between different datasets(in the same zpool). Is reflink not supported between different datasets?

**Reply by IvanVolosyuk** [reply]
It is supported in ZFS 2.2, but you may have to do ```zpool upgrade [pool]``` to enable the feature flag.
You can check if the feature is ```enabled``` using:
```
zpool get feature@block_cloning [pool]
```
There may be also be additional limitations for encrypted pools.

**Reply by rincebrain** [reply]
`=auto` should never say that even if the feature is disabled, so if you try it again to confirm that's exactly the flag you used, it's a bug, I think. `=always` would, however, never work cross dataset because Linux hardcodes a check about that.

**Reply by shivabohemian** [reply]
Sorry. I used `cp --reflink=always src dst` in order to always use the shared data.
The `feature@block_cloning` is enabled and it works very well(cp fast)  when using `cp --reflink=always` in the same dataset. 
`cp --reflink=always` would never work cross dataset indeed because of the Linux hardcodes. So I use `cp --reflink=auto src dst` to cp files cross dataset, it's much slower than in the same dataset. 
It seems like expected behavior, right?

**Reply by rincebrain** [reply]
`cp --reflink=auto` should still block clone cross-dataset on the same pool unless encryption is involved, so no slower.

**Reply by shivabohemian** [reply]
You can take a look at my test below, which is two unencrypted datasets
```
root@testserver:/test/dataset1# zfs list
NAME            USED  AVAIL  REFER  MOUNTPOINT
test           5.98G  7.12T  4.00G  /test
test/dataset1  1.97G  7.12T  1.97G  /test/dataset1
test/dataset2    96K  7.12T    96K  /test/dataset2
root@testserver:/test/dataset1# du -sh *
2.0G	dataset.zip
root@testserver:/test/dataset1# time cp --reflink=auto dataset.zip dataset2.zip

real	0m0.209s
user	0m0.004s
sys	0m0.202s
root@testserver:/test/dataset1# du -sh *
1.9G	dataset2.zip
2.0G	dataset.zip
root@testserver:/test/dataset1# time cp --reflink=auto dataset.zip /test/dataset2/dataset2.zip

real	0m18.424s
user	0m0.099s
sys	0m8.059s
root@testserver:/test/dataset1# zfs get encryption test
NAME  PROPERTY    VALUE        SOURCE
test  encryption  off          default
root@testserver:/test/dataset1# zfs get encryption test/dataset1
NAME           PROPERTY    VALUE        SOURCE
test/dataset1  encryption  off          default
root@testserver:/test/dataset1# zfs get encryption test/dataset2
NAME           PROPERTY    VALUE        SOURCE
test/dataset2  encryption  off          default
```

**Reply by rincebrain** [reply]
I suppose it also depends on the kernel version, if it's too old, it's not going to work either.

**Reply by shivabohemian** [reply]
All right. The kernel version I use is 5.10.

**Reply by hdiethelm** [reply]
Anyone has an idea after which versions this is sported?
I am using debian bookworm with backports:
coreutils 9.1-1
kernel 6.12.9
zfs 2.2.7
And I just get operation not supported when using always / standard copy when using auto.

**Reply by jonmchan** [reply]
Proxmox VE 8.3.5 does not seem to support this yet.

```
# zfs --version
zfs-2.2.7-pve1
zfs-kmod-2.2.7-pve1

# pveversion
pve-manager/8.3.5/dac3aa88bac3f300 (running kernel: 6.8.12-8-pve)

# zfs get encryption spool/storage/video-content
NAME                         PROPERTY    VALUE        SOURCE
spool/storage/video-content  encryption  off          default

# zfs get encryption spool/storage
NAME           PROPERTY    VALUE        SOURCE
spool/storage  encryption  off          default

# ls -lha /spool/storage/MyLargeFile.7z
-rwxr-xr-x 1 3000000 3000000 6.0G Dec  6  2013 /spool/storage/MyLargeFile.7z

# time cp --reflink=auto /spool/storage/MyLargeFile.7z /spool/storage/video-content/

real    0m21.519s
user    0m0.033s
sys     0m3.958s
```

**Reply by melianmiko** [reply]
Proxmox 9.0, `--reflink=auto` causes memory leak and machine hang...

**Reply by 1JorgeB** [reply]
While testing for #17629, which is not a ZFS bug, I've found that sometimes I can reflink across datasets, other times not. 

In this case, I'm happy it worked like that, because when cp reverts to a normal copy, it's when that bug is triggered; when it reflinks it doesn't trigger the bug.

I just don't understand why reflinks sometimes work across datasets and not others, everything else is the same, both datasets have the same (default) settings, just by recreating the pool it can sometimes allow it and other times not.

kernel 6.12.24
openzfs 2.3.1

```
tank       35.1G   201G    96K  /mnt/tank
tank/one   7.05G   201G  7.05G  /mnt/tank/one
tank/two   28.0G   201G  28.0G  /mnt/tank/two
```

```
root@Test5:~# time cp --debug /mnt/tank/one/new.iso /mnt/tank/two/new.iso
'/mnt/tank/one/new.iso' -> '/mnt/tank/two/new.iso'
copy offload: yes, reflink: unsupported, sparse detection: SEEK_HOLE

real    0m0.282s
user    0m0.001s
sys     0m0.278s
```

```
root@Test5:~# time cp --debug /mnt/tank/one/new.iso /mnt/tank/two/new.iso 
'/mnt/tank/one/new.iso' -> '/mnt/tank/two/new.iso'
copy offload: unsupported, reflink: unsupported, sparse detection: SEEK_HOLE

real    0m22.561s
user    0m0.014s
sys     0m2.105s
```

**Reply by robn** [reply]
This is a little strange. `copy offload: unsupported` means `copy_file_range()` failed, and that shouldn't happen really, unless I've forgotten some detail (possible; it's all a bit weird down here).

Can you retry those commands under strace, that is: `strace -e trace=ioctl,copy_file_range cp --debug ...`. Here, for same dataset, I get a proper clone ("reflink") first time:

```
$ strace -e trace=ioctl,copy_file_range cp --debug memtest86-usb.img copy.img
'memtest86-usb.img' -> 'copy.img'
ioctl(4, BTRFS_IOC_CLONE or FICLONE, 3) = 0
copy offload: unknown, reflink: yes, sparse detection: unknown
+++ exited with 0 +++
```

And for cross-dataset, the clone fails as expected, and I get a fallback to offload (which may or may not result in a clone, by design):

```
$ strace -e trace=ioctl,copy_file_range cp --debug memtest86-usb.img /dump/copy.img
'memtest86-usb.img' -> '/dump/copy.img'
ioctl(4, BTRFS_IOC_CLONE or FICLONE, 3) = -1 EXDEV (Invalid cross-device link)
copy_file_range(3, NULL, 4, NULL, 131072, 0) = 131072
copy_file_range(3, NULL, 4, NULL, 8126464, 0) = 8126464
copy_file_range(3, NULL, 4, NULL, 8126464, 0) = 8126464
copy_file_range(3, NULL, 4, NULL, 131072, 0) = 131072
copy offload: yes, reflink: unsupported, sparse detection: SEEK_HOLE
+++ exited with 0 +++
```

So if you've got a failing `copy_file_range()`, that's curious to me.

Of course, I don't trust the broken glibcs at all here, if that's what you have. Also coreutils version matters, check `cp --version`.

For reference, my versions are:

```
$ uname -sr
Linux 6.12.41+deb13-amd64

$ zfs version
zfs-2.3.4-lucy1
zfs-kmod-2.3.4-lucy1

$ cp --version
cp (GNU coreutils) 9.7

$ dpkg -s libc6 | grep Version
Version: 2.41-12
```

**Reply by 1JorgeB** [reply]
> Of course, I don't trust the broken glibcs at all here

You may be correct, and that is what is causing the issue, I've been testing with glibc 2.41 and 2.42, and with both I see the described behavior: cross-dataset reflinks sometimes work and sometimes don't, under the exact same circumstances.

But I now had the chance to retest with glibc 2.42 with the [patch](https://sourceware.org/git/?p=glibc.git;a=blobdiff;f=nptl/cancellation.c;h=bed0383a23f0ac1d587ff4296c0bdd3db5d568a8;hp=156e63dcf0dc9d9c858deb15c5aca4f00c7c190a;hb=7107bebf19286f42dcb0a97581137a5893c16206;hpb=cf91048748408f81566ead5b9498744100aa589c) to fix #17629 and cross-dataset reflinks always work, at least they have so far.

In case there's still interest, this is the info requested:

```
$ cp --version
cp (GNU coreutils) 9.7
```

Don't have dpgk in this system, but glibc in this case was 2.41 (without the patch), example output from when cross-dataset reflink doesn't work (and the file gets corrupted):

```
# time strace -e trace=ioctl,copy_file_range cp --debug /mnt/tank/one/new.iso /mnt/tank/two/new.iso
'/mnt/tank/one/new.iso' -> '/mnt/tank/two/new.iso'
ioctl(4, BTRFS_IOC_CLONE or FICLONE, 3) = -1 EXDEV (Invalid cross-device link)
copy_file_range(3, NULL, 4, NULL, 131072, 0) = 131072
copy_file_range(3, NULL, 4, NULL, 1703936, 0) = 1703936
copy_file_range(3, NULL, 4, NULL, 2359296, 0) = 2359296
copy_file_range(3, NULL, 4, NULL, 3407872, 0) = 3407872
copy_file_range(3, NULL, 4, NULL, 2490368, 0) = 2490368
copy_file_range(3, NULL, 4, NULL, 3744989184, 0) = 3744989184
copy_file_range(3, NULL, 4, NULL, 2359296, 0) = 2359296
copy_file_range(3, NULL, 4, NULL, 3891134464, 0) = 3891134464
copy offload: unsupported, reflink: unsupported, sparse detection: SEEK_HOLE
+++ exited with 0 +++

real    0m11.107s
user    0m0.076s
sys     0m2.337s
```
Output with the patched glibc 2.42, reflink always works so far, and when it reflinks, there's no corruption.

```
# time strace -e trace=ioctl,copy_file_range cp --debug /mnt/tank/one/new.iso /mnt/tank/two/new2.iso
'/mnt/tank/one/new.iso' -> '/mnt/tank/two/new2.iso'
ioctl(4, BTRFS_IOC_CLONE or FICLONE, 3) = -1 EXDEV (Invalid cross-device link)
copy_file_range(3, NULL, 4, NULL, 131072, 0) = 131072
copy_file_range(3, NULL, 4, NULL, 1703936, 0) = 1703936
copy_file_range(3, NULL, 4, NULL, 2359296, 0) = 2359296
copy_file_range(3, NULL, 4, NULL, 3407872, 0) = 3407872
copy_file_range(3, NULL, 4, NULL, 2490368, 0) = 2490368
copy_file_range(3, NULL, 4, NULL, 3744989184, 0) = 3744989184
copy_file_range(3, NULL, 4, NULL, 2359296, 0) = 2359296
copy_file_range(3, NULL, 4, NULL, 3891134464, 0) = 3891134464
copy_file_range(3, NULL, 4, NULL, 1179648, 0) = 1179648
copy_file_range(3, NULL, 4, NULL, 35897344, 0) = 35897344
copy offload: yes, reflink: unsupported, sparse detection: SEEK_HOLE
+++ exited with 0 +++

real    0m0.419s
user    0m0.004s
sys     0m0.358s
```

When it doesn't do a reflink with the broken glibc, some files with holes get corrupted; you don't even need to do a checksum comparison, since it grows in size:

```
# ls -l /mnt/tank/one/
total 7384916
-rwxrw-rw- 1 root root 7688142848 Aug 21 17:42 new.iso*

# ls -l /mnt/tank/two/
total 18380302
-rwxrw-rw- 1 root root 11433132032 Aug 22 11:27 new.iso*
-rwxrw-rw- 1 root root  7688142848 Aug 22 11:32 new2.iso*

```

So I guess this strange cross-dataset behavior was all being caused by the broken glibc.

P.S. I'm testing this with Unraid; the patched 2.42 glicb comes with a newer release that also includes newer kernel and Openzfs, don't think that is the issue, but for completion, that release uses

```
kernel 6.12.42
openzfs 2.3.3-1
cp (GNU coreutils) 9.7
glibc 2.42 with the patch
```

**Reply by robn** [reply]
Yeah, this is all the glibc breakage. "Files with holes" is a big giveaway; it's not _necessary_, but its an easy way for `copy_file_range()` to advance over 2G in a single call, which is where the broken glibc would misinterpret the return from the kernel. So yeah, nothing to see here.

Incidentally, none of this is cloning across datasets. `copy_file_range()` when it works is not "clone the file", but just "make another one, I don't care how". It's not really relevant to this discussion, it's just that "reflink" is kind of ambiguous so I just wanted to mention it.

**Reply by 1JorgeB** [reply]
> "Files with holes" is a big giveaway

Right, since my intention was to test the glibc bug, I was always using a specific file with holes that triggers that bug, but now that you mention it, I retested with a file that doesn't trigger the bug, and a reflink is always done with the broken glibc utils, confirming that's the problem.

Still strange to me that using the same file with holes and the same pool/dataset setup with the affected glibc versions, sometimes made a reflink, other times not, but it doesn't really matter now since it is a glibc problem and it's now resolved.

Thanks for your time, and sorry since this wasn't really related to zfs.

**Edit to add**: Like mentioned, doesn't really matter, but it was bothering me, I've found that sometimes being able to make a reflink for the same file, other times not, is not related to the pool, but possibly to how that file is copied/written/aligned.

I'm not sure which yet, but I can show the same file showing both behaviors in the same pool: 
cp creates a reflink, no corruption; cp reverts to a regular copy, corruption. This is with the broken glibc utils, of course.

<img width="751" height="590" alt="image" src="https://github.com/user-attachments/assets/145aa21f-134b-42f2-81fd-f43a885db86e" />

**Reply by hdiethelm** [reply]
> Anyone has an idea after which versions this is sported? I am using debian bookworm with backports: coreutils 9.1-1 kernel 6.12.9 zfs 2.2.7 And I just get operation not supported when using always / standard copy when using auto.

Just an update from my side:
Debian Trixie
zfs-2.3.2
cp (GNU coreutils) 9.7

It just works, no arguments needed. I just figured that out when I was moving files in Gnome which should have taken a long time across datasets but it was done in seconds. Great job by someone.

cp below takes ~1s for 32GB:
zpool get all tank1 | grep bclone
tank1  bcloneused                     33.8G                          -
tank1  bclonesaved                    59.8G                          -
tank1  bcloneratio                    2.77x 

cp --debug test123 test1234
'test123' -> 'test1234'
copy offload: unknown, reflink: yes, sparse detection: unknown

zpool get all tank1 | grep bclone
tank1  bcloneused                     33.8G                          -
tank1  bclonesaved                    85.8G                          -
tank1  bcloneratio                    3.54x

And accross datasets:
cp --debug test123 ../testcp/
'test123' -> '../testcp/test123'
copy offload: unsupported, reflink: unsupported, sparse detection: SEEK_HOLE

zpool get all tank1 | grep bclone
tank1  bcloneused                     33.8G                          -
tank1  bclonesaved                    112G                           -
tank1  bcloneratio                    4.31x   

By the way: Don't use rsync, that does a old style copy allocating new blocks.
